src/harness/reference_models/fss_purge/fss_purge.py

```diff
@@ -0,0 +1,332 @@
+    fss_record: A FSS record dictionary.
+  Returns:
+    List of CBSDs lying within 40 KMs of FSS.
+  """
+
+  neighbouring_cbsds = []
+  for cbsd in cbsd_list:
+    distance_km, _, _ = vincenty.GeodesicDistanceBearing(
+        fss_record['deploymentParam'][0]['installationParam']['latitude'],
+        fss_record['deploymentParam'][0]['installationParam']['longitude'],
+        cbsd['registrationRequest']['installationParam']['latitude'],
+        cbsd['registrationRequest']['installationParam']['longitude'])
+
+    if distance_km <= 40:
+        neighbouring_cbsds.append(cbsd)
+  return neighbouring_cbsds
+
+def db2Linear(x):
+  """ Convert dB to linear scale. """
+  return 10**(x/10.)
+
+def performPurge(cbsds, fss_entity, fss_oobe_margin):
+  """ Implementation of algorithm described in R2-SGN-29.
+
+  The list of grants to be purged based out of band emission are identifies and the grants
+  are removed from the CBSDs. If a CBSDs has no grants then the CBSD is removed from the list.
+
+  Args:
+    cbsds: List of CbsdData objects that are in the neighbourhood to an FSS.
+    fss_entity: The FssProtectionPoint tuple of the FSS for which the purge is to be performed.
+    fss_oobe_margin: OOBE Margin value to be considered for FSS purge list calculation.
+  """
+
+  # Calculate the OOBE interference value for each CBSD based on the grant with the highest frequency.
+  calculateOobeInterference(cbsds, fss_entity)
+
+  # Sort the CBSDs based on the calculated interference value.
+  sorted_cbsds = sorted(cbsds, key=lambda cbsd: cbsd['oobe_interference'])
+  # Calculate the threshold value in dBm for the OOBE interference
+  oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/ONE_MHZ) - fss_oobe_margin)
+
+  # Find the largest index such that aggregation of the interferences do not cross 
+  # the OOBE threshold
+  oobe_interferences = [db2Linear(cbsd['oobe_interference']) for cbsd in sorted_cbsds]
+  cumulated_interference = np.cumsum(oobe_interferences)
+  index = np.searchsorted(cumulated_interference, db2Linear(oobe_threshold_value))
+  cbsds_to_purge = sorted_cbsds[index:]
+  sorted_cbsds = sorted_cbsds[:index]
+
+  # Remove the grants from the CBSDs that are in the purge list.
+  if cbsds_to_purge:
+    for index, cbsd in enumerate(cbsds_to_purge):
+      # Get the grant that was used for the OOBE interference calculation, that
+      # is the grant with highest highFrequency.
+      grant_with_max_high_freq = max(cbsd['grantRequests'],\
+            key=lambda x: x['operationParam']['operationFrequencyRange']['highFrequency'])
+
+      # Get all the grants that share the same emission mask as the one that 
+      # was used for OOBE calculation
+      if (grant_with_max_high_freq['operationParam']['operationFrequencyRange']
+          ['highFrequency'] > 3690000000):
+
+        # Adding the -13 emission mask grants to the purge list 
+        
+        grants_to_purge = [grant_request for grant_request in cbsd['grantRequests'] 

```

Comment:
If `cbsd` is a `CbsdData` object from SAS-SAS, then you need to use `grants` not `grantRequests` here.

Reply:
Updated